src/backend/distributed/utils/attribute.c

```diff
@@ -0,0 +1,619 @@
+			{
+				break;
+			}
+
+			tenantEnd++;
+		}
+		tenantEnd--;
+
+		colocationGroupId = 0;
+		while (*tenantEnd != ',')
+		{
+			colocationGroupId *= 10;
+			colocationGroupId += *tenantEnd - '0';
+			tenantEnd--;
+		}
+
+		int t = colocationGroupId;
+		colocationGroupId = 0;
+		while (t)
+		{
+			colocationGroupId *= 10;
+			colocationGroupId += t % 10;
+			t /= 10;
+		}
+
+		/* hack to get a clean copy of the tenant id string */
+		char tenantEndTmp = *tenantEnd;
+		*tenantEnd = '\0';
+		tenantId = pstrdup(tenantId);
+		*tenantEnd = tenantEndTmp;
+
+		if (MultiTenantMonitoringLogLevel != CITUS_LOG_LEVEL_OFF)
+		{
+			ereport(NOTICE, (errmsg("attributing query to tenant: %s",
+									quote_literal_cstr(tenantId))));
+		}
+
+		/*	attributeToTenant = (char *) malloc(strlen(tenantId)); */
+		strcpy_s(attributeToTenant, sizeof(attributeToTenant), tenantId);
+	}
+	else
+	{
+		/*Assert(attributeToTenant == NULL); */
+	}
+
+	/*DetachSegment(); */
+
+	attributeToTenantStart = clock();
+}
+
+
+/*
+ * AnnotateQuery annotates the query with tenant attributes.
+ */
+char *
+AnnotateQuery(char *queryString, char *partitionColumn, int colocationId)
+{
+	if (partitionColumn == NULL)
+	{
+		return queryString;
+	}
+	StringInfo newQuery = makeStringInfo();
+	appendStringInfo(newQuery, ATTRIBUTE_STRING_FORMAT, partitionColumn, colocationId);
+
+	appendStringInfoString(newQuery, queryString);
+
+	return newQuery->data;
+}
+
+
+void
+CitusAttributeToEnd(QueryDesc *queryDesc)
+{
+	/*
+	 * At the end of the Executor is the last moment we have to attribute the previous
+	 * attribution to a tenant, if applicable
+	 */
+	AttributeMetricsIfApplicable();
+
+	/* now call in to the previously installed hook, or the standard implementation */
+	if (prev_ExecutorEnd)
+	{
+		prev_ExecutorEnd(queryDesc);
+	}
+	else
+	{
+		standard_ExecutorEnd(queryDesc);
+	}
+}
+
+
+/*
+ * AttributeMetricsIfApplicable updates the metrics for current tenant's statistics
+ */
+static void
+AttributeMetricsIfApplicable()
+{
+	if (strcmp(attributeToTenant, "") != 0)
+	{
+		clock_t end = { 0 };
+
+		end = clock();
+		time_t queryTime = time(0);
+		double cpu_time_used = ((double) (end - attributeToTenantStart)) / CLOCKS_PER_SEC;
+
+		if (MultiTenantMonitoringLogLevel != CITUS_LOG_LEVEL_OFF)
+		{
+			ereport(NOTICE, (errmsg("attribute cpu counter (%f) to tenant: %s",
+									cpu_time_used, attributeToTenant)));
+		}
+
+		MultiTenantMonitor *monitor = GetMultiTenantMonitor();
+
+		LWLockAcquire(&monitor->lock, LW_SHARED);
+
+		monitor->periodStart = monitor->periodStart +
+							   ((queryTime - monitor->periodStart) /
+								CitusStatsTenantsPeriod) *
+							   CitusStatsTenantsPeriod;
+
+		int tenantIndex = FindTenantStats(monitor);
+
+		if (tenantIndex == -1)
+		{
+			tenantIndex = CreateTenantStats(monitor);
+		}
+		TenantStats *tenantStats = &monitor->tenants[tenantIndex];
+
+		LWLockAcquire(&tenantStats->lock, LW_EXCLUSIVE);
```

Comment:
could this cause contention?

Reply:
Had offline discussion. We are keeping this for now, until we run some benchmarks.